invoices/invoiceregistry.go

```diff
@@ -880,11 +839,23 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
 
 	switch invoiceHtlc.State {
 	case channeldb.HtlcStateCanceled:
-		return NewFailureResolution(
-			circuitKey, acceptHeight, result,
-		), nil
+		failureResolution := resolution.(*HtlcFailResolution)
```

Comment:
Unfortunately there just isn't good separation of duties between the callback and `NotifyExitHopHtlc` at the moment. Eg, before this change we're just assuming that the resolution result that we make the `FailureResolution` with if a result related to a failure because the htlc state is cancelled. This change more strictly controls that, but we're still operating under the assumption that our resolution matches our current status. I want to come back in another PR and refactor further to more cleanly separate the two.

Reply:
Updated to rather switch on resolution type after offline discussion if you want to have another look :)